rust/src/meas/randomized_response/mod.tex

```diff
@@ -0,0 +1,250 @@
+\newtheorem{hope}{Hope}
+\newtheorem{warning}{Warning}
+\newtheorem{problem}{Problem}
+\newtheorem{fear}{Fear}
+\newtheorem{question}{Question}
+
+\title{Privacy Proofs for OpenDP: Binary Randomized Response Measurement}
+\author{Vicki Xu, Hanwen Zhang, Zachary Ratliff}
+\date{Summer 2022}
+\begin{document}
+
+\maketitle
+\tableofcontents
+
+%\section{Measurement Description}
+%In binary randomized response, the participant lies with a probability $< \frac{1}{2}$ and tells the truth for the rest of the time. 
+%In Python and in Rust, when booleans are cast to integers, $1$ represents true, $0$ false. 
+%In the function defined line \ref{line:fn}, the true value, \emph{arg}, is taken as input. A boolean is generated in line \ref{line:coinflip} using \texttt{sample\_bernoulli}, which takes in \emph{prob} and \emph{constant\_time} as inputs. As we see from part (a), this means a value of $1$ is generated \emph{prob} percent of the time, and $0$ is generated $1 -$\emph{prob} percent of the time. This is the ``coinflip" step, wherein the computer decides whether to return the true value or the random response. So with probability \emph{prob}, the true value, \emph{arg}, is returned. With probability $1 -$\emph{prob}, the lie, which in the case of the boolean is the opposite (hence $1 - $\emph{arg}), is returned. 
+
+\section{Algorithm Implementation}
+\subsection{Code in Rust}
+
+The current OpenDP library contains the \texttt{make\_randomized\_response\_bool} function implementing the binary randomized response measurement. This is defined in lines 28-53 of the file \texttt{mod.rs} in the Git repository \url{https://github.com/opendp/opendp/blob/main/rust/src/meas/randomized_response/mod.rs#L28-L53}
+
+In \texttt{make\_randomized\_response\_bool}, which accepts a parameter \texttt{prob} of type \texttt{Q} and a parameter \texttt{constant\_time} of type \texttt{bool}, the function takes in a boolean data point \texttt{arg} and returns the truthful value \texttt{arg} with probability \texttt{prob} and the untruthful value $\texttt{!arg}$ with probability $1 - \texttt{prob}$ (in constant time, if the flag is turned on). 
+
+\subsection{Pseudo Code in Python}
+We present a simplified Python-like pseudocode of the Rust implementation below. The necessary definitions for the pseudocode can be found at \href{https://github.com/opendp/whitepapers/blob/pseudocode-defns/pseudocode-defns/pseudocode_defns.pdf}{``List of definitions used in the pseudocode"}. % need to fix
+
+\emph{The use of \texttt{code}-style parameters in the preconditions section below (for example, \texttt{input\_domain}) means that this information should be passed along to the \texttt{Measurements} constructor.}
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{prob} must be of type \texttt{Q}
```

Comment:
I was unable to find the definition of type Q in the pseudocode defs doc (though it is used a lot in the metrics section of that doc)

Reply:
@Shoeboxam what exactly is the difference between types T and Q? are they both "generic types"?

Reply:
They are both generic types. These are just naming conventions:

```
- D for Domain
- M for Metric or Measure
- T for D::Carrier
- Q for M::Distance
```
The above is ordered by precedence, and we don't introduce redundant types. If there is a constructor where T == Q, then it will only have one generic named T (because T has higher precedence), and a generic Q won't exist in the code. 

If T != Q, then you'll see both generics in use, like in `make_randomized_response`.

Similarly if we have a generic D, then there won't also be a generic T because T is redundant with D::Carrier. So D has higher precedence.

Additionally:
```
- *I for Input. Like TI or MI
- *O for output
- **A for Atom. Like TIA or DA
```